### DISCUSSION TITLE
inference - many tracks for 2 instances

### DISCUSSION THREAD

**Discussion by akshitasax** [talmolab/sleap#771] [General]
Hi,

I work with 12 hour long video files. After running inference on the entire video, I usually get around 200 tracks broken up throughout the video. I usually have only 2 instances. Proofreading this would not be feasible. What can I do to help this? This happens when using default settings. I have also tried doing cull to target no. of instances, but just that didn't help much. 

I am also having an issue with the frequency of switched tracks. In some parts of the video it is almost as if the tracking is random. Can I do something to help this too?

Thank you

![Screenshot from 2022-05-31 18-03-48](https://user-images.githubusercontent.com/104799583/171314476-a1e83740-4003-41bc-ac55-432fa94c2ab5.png)
**Screenshot 1: Default Inference Pipeline settings.**

**Comment by roomrys** [comment]
Hi @akshitasax,

Have you tried using a Kalman Filter during inference? The [KF](https://sleap.ai/api/sleap.nn.tracker.kalman.html#:~:text=The%20Kalman%20filters,up%20the%20filters.) should help minimize switching. 

Also what [matching method](https://sleap.ai/develop/guides/proofreading.html#tracking-methods:~:text=Once%20SLEAP%20has,Hungarian%E2%80%9D%20matching.) are you using? Using Greedy matching will assign tracks based on the cost of matching in the current frame only while Hungarian matching will take into account the long-term cost across the all frames to predict on.

![image](https://user-images.githubusercontent.com/38435167/171465595-21418759-2f72-4423-89bf-70b576434ad7.png)
**Fig 1: Location of parameters in the Inference Pipeline that could help minimize track switching.**

Let us know if any of the above suggestions helped in your case and/or if you have more questions.

Tagging similar questions here: #714 ([helpful comment](https://github.com/talmolab/sleap/issues/714#issuecomment-1115460151)) and [How to get continuous tracking during inference](https://github.com/talmolab/sleap/discussions/741#discussion-4070525) (which ambiguously concludes that KF was the problem).

Thanks,
Liezl

**Reply by akshitasax** [reply]
I just tried using the flow method, while targeting 2 instances, using hungarian matching, and enabling Kalman filters. An error message popped up. I then tried simple method with hungarian, KF and culling to target instance, and the same message popped up. Attached is a screenshot showing the message and the command line terminal at that time. The second screenshot shows the settings I used. Just yesterday I was able to run inference using simple method, with hungarian matching, culling to target instances and connecting single track breaks (so, all the same settings except KF). Do you think this is a GPU space problem?

![Screenshot from 2022-06-01 12-37-21](https://user-images.githubusercontent.com/104799583/171488182-4d0525bb-2a89-4598-8828-b736ced4d8b7.png)

![Screenshot from 2022-06-01 12-37-38](https://user-images.githubusercontent.com/104799583/171488829-029f17cb-b80d-47a4-ab5e-3c0fc39cbf5e.png)

**Reply by akshitasax** [reply]
Oh I think I understood the problem - I think "nodes to use for tracking" should be 0 instead of 0,1,2 since I only have 1 node. It is running now.

**Reply by akshitasax** [reply]
Okay I ran inference using these settings (screenshot attached). However there were again around 200 tracks, and many frames had only one track even though both instances have been predicted. (screenshot attached of one such frame. also shown is the seekbar showing the many tracks throughout the video). Anything else I can try to help this? Yesterday with just culling to target no. of instances and selecting "connect single track breaks", but without KF, I was able to run inference and obtain no more than 2 tracks throughout the video. The switching was still a problem, so today I tried it with the KF enabled. But then this was the result. 

![Screenshot from 2022-06-01 15-02-00](https://user-images.githubusercontent.com/104799583/171509808-d0b3cd90-a6c7-478f-a254-a7805e269e8b.png)
**Fig 2: Inference settings - KF enabled.**

![Screenshot from 2022-06-01 15-03-40](https://user-images.githubusercontent.com/104799583/171509862-8a36e17e-7bb3-4988-b8a9-c1969e02dabc.png)
**Screenshot 2: frame that has 2 predicted instances, but only 1 track. Seekbar showing all the different tracks in the video.**

**Reply by roomrys** [reply]
@akshitasax Hmm... ok so in the first screenshot you sent it looks like the tracks were relatively stable at the beginning of the video whereas now the tracks are doing even more switching. The KF uses a sample of tracks as input to initialize the filter (namely the number of frames specified in the Training/Inference pipeline). However, it seems that the method of matching has messed up the initial stability of our tracks and may have poorly initialized the KF - although this shouldn't really matter as KF should converge regardless of the initialization. 

Lets isolate the methods. Are you able to run four separate Inference Pipelines:
1. Using Hungarian matching no KF
2. Using Hungarian matching with KF (with [**two**](https://github.com/talmolab/sleap/discussions/771#discussioncomment-2874631) frames)
3. Using Greedy matching with KF (with [**two** ](https://github.com/talmolab/sleap/discussions/771#discussioncomment-2874631)frames)
4. Using Greedy matching with KF (and possibly a larger set of initialization frames - up to the point in the first screenshot where the tracks are still relatively stable)

Lets see how that works.

Thanks,
Liezl

**Reply by akshitasax** [reply]
by single frame do you mean the number for "enable filters after initial frames"? Make that 1 instead of 10 for pipeline 2 and 3?

**Reply by roomrys** [reply]
@akshitasax Yes, exactly.

**Reply by akshitasax** [reply]
does the similarity method (centroid, iou, or instance) matter?

**Reply by akshitasax** [reply]
So I tried the first piepline, and I did get a manageable number of tracks, (it was more than 2 but probably still able to be proofread), and the switching was reduced too. (attached screenshot).
![Screenshot from 2022-06-02 14-38-02](https://user-images.githubusercontent.com/104799583/171743102-8c1df654-2711-4f40-8aef-1b75deec0ce0.png)


When I tried to run the second pipeline it gave me an error. I am not sure how to fix this. (screenshot attached of the inference pipeline settings and the command line terminal at the time the error happened). I'm guessing it has something to do with the line "could not broadcast input array from shape (4,2) into shape (4,1)" but I am not sure what this refers to.
![Screenshot from 2022-06-02 14-35-47](https://user-images.githubusercontent.com/104799583/171743086-6415b51a-a93d-402b-8a21-caa6b8c33b82.png)

(edited) Okay so I tried running the same settings but with KF frames as 10 instead of 1, and it ran. Why do you think it doesn't work with 1 frame?

**Reply by roomrys** [reply]
@akshitasax 
Much better results for Hungarian matching!

I apologize, taking a closer look at our KF implementation, it seems we use pykalman's implementation of the Expectation Maximization algorithm (which needs multiple frames as input) to estimate the parameters used in the KF. 

I wanted to see how the KF fairs with a single frame because the best initialization to the KF is the past observation (assuming that the observation is correct/accurate). I tested to make sure the KF works with 2 frames - lets test with 2 initialization frames on your training set. I updated the Pipelines above.

**Future GUI upgrade should put limits on user inputs that will break code.**

Thanks,
Liezl

**Reply by akshitasax** [reply]
Hi @roomrys 

I tried all 4 inference pipelines you had suggested and the best predictions were obtained using hungarian matching without KF. It gave me only two tracks, while all other pipelines resulted in close to 200. I also changed my skeleton to include three nodes (head, thorax and abdomen) and that reduced the frequency of identity switching considerably. However, proofreading was still required as there were some switches. For proofreading I've been seeking through the entire video in order to find frames where switching occurs. Currently this takes about 2-2.5 hours for two instances in a 12-hour video. Do you have any suggestions as to how I can reduce this proofreading time further? This time will probably also increase as I increase the number of flies I want to track. Have others tried to use SLEAP to track such large videos, and what does their proofreading time look like? 

Thank you,
Akshita

**Reply by roomrys** [reply]
Hi @akshitasax,

Thanks for following up with which pipeline worked best! Noted that Kalman Filtering is not producing the expected outcome...

[Proofreading](https://sleap.ai/develop/guides/proofreading.html#id1:~:text=One%20strategy%20is%20to%20set%20the%20trail%20length%20to%20a%20number%20greater%20than%200%20(e.g.%2050)%20and%20jump%20through%20the%20predictions%20using%20the%20frame%20next%20large%20step%20hotkey.%20It%E2%80%99s%20usually%20possible%20to%20see%20identity%20swaps%20by%20looking%20at%20the%20shape%20of%20the%20track%20trails) can be a tedious process, but to speed it up, you can set the trail length (View > Trail Length) to 100 or 150 and use the Frame Next Large Step keyboard shortcut (Help > Keyboard Shortcuts > Frame Next Large Step) to step through the frames in a large step (currently set to 100 frames/step). The length and shape of the trails should help visualize where switching occurs.

Thanks,
Liezl
